### ISSUE TITLE
dict2inst should ensure type correctness

### ISSUE THREAD

**Issue by afonsolage** [godotengine/godot#48169]
**Godot version:**
3.3.stable

**OS/device including version:**

Ubuntu 20.04, GLES3

**Issue description:**
When converting from json parsed dictionary to instance, `dict2inst`, does not keep the correct type. Since JSON always uses float, all int properties are set as float, even if using typed gdscript.

```gdscript
extends Spatial

enum Test {A, B, C}

var i: int = 1

func _ready():
	print_match(i)
	var another = dict2inst(parse_json(to_json(inst2dict(self))))
	print_match(another.i)


func print_match(value):
	match value:
		Test.A:
			print("A")
		Test.B:
			print("B")
		Test.C:
			print("C")
		_:
			print("None")
```

The above code should print "B" even after the parsing, but instead, it fails on second match pass:
```
B
None
```

As a workaround, I made the following function, to fix the deserialized object:
```gdscript
func fix_ints(obj: Object) -> void:
	for prop in obj.get_property_list():
		if prop.type == TYPE_INT:
			obj.set(prop.name, obj.get(prop.name) as int)
```

**Steps to reproduce:**
1. Open the project
2. Run the project
3. See results

**Minimal reproduction project:**
[project.zip](https://github.com/godotengine/godot/files/6370854/project.zip)

**Comment by akien-mga** [comment]
There are no integers in JSON, there's no way to preserve this information.

If you want to keep exact type info, use ConfigFile instead of JSON.

**Comment by afonsolage** [comment]
Yes, I know this is because of how JSON works, but `dict2inst` show try to cast the type or at least, throw an error, because having an object with a property of `TYPE_INT` but with a float value, doesn't seems right.

**Comment by akien-mga** [comment]
Well in your example, there's no indication that `another.i` should be an int. It's whatever can be inferred from the JSON, it has no link to the member `i` which had been converted to JSON.

**Comment by Xrayez** [comment]
I have a PR which can solve the issue with JSON, see #33241, but for it to be merged it needs an approval from the core developers.

**Comment by afonsolage** [comment]
> Well in your example, there's no indication that `another.i` should be an int. It's whatever can be inferred from the JSON, it has no link to the member `i` which had been converted to JSON.

Isn't `dict2inst` suppose to do this link? Since it converts a dictionary, which has a link to the script, to the object.

Anyways, I made a small change to the test script, using a subclass and casting the type:

```gdscript
extends Spatial

enum Test {A, B, C}

class Sub:
	var i: int = 1

func _ready():
	var sub := Sub.new()
	for info in sub.get_property_list():
		if info.name == "i":
			print(info)
			
	print_match(sub.i)
	var another = dict2inst(parse_json(to_json(inst2dict(sub)))) as Sub
	print_match(another.i)
	
	for info in another.get_property_list():
		if info.name == "i":
			print(info)


func print_match(value):
	match value:
		Test.A:
			print("A")
		Test.B:
			print("B")
		Test.C:
			print("C")
		_:
			print("None")

```

This prints the following:
```
{class_name:, hint:0, hint_string:, name:i, type:2, usage:8192}
B
None
{class_name:, hint:0, hint_string:, name:i, type:2, usage:8192}
```

So the `dict2inst` converts the dictionary to instance without problems, but it ends with a int property that has a float. For consistence, I think `dict2inst` should ensure the type when casting back.

**Comment by akien-mga** [comment]
Thanks, I understand the issue better now. The JSON number vs float/int issue is not the real issue, it's just a way to expose it, so let's simplify:

```gdscript
extends Spatial

class Sub:
	var i: int = 1

func _ready():
	var dict = { "@path":"res://Spatial.gd", "@subpath":"Sub", "i":42.0 }
	var inst = dict2inst(dict)
	print(inst is Sub)  # True
	print(typeof(inst.i) == TYPE_INT)  # False, expected is True
```

Since `dict2inst` does create an instance of the object defined by `@path` and `@subpath`, one could expect that it would either do type coercion or raise an error if the dict contains different types.

Instead, it silently assigns a `float` value to a member with `int` type hint.

**Comment by HolonProduction** [comment]
`dict_to_inst` is deprecated by now, so I don't think a change like this makes sense.
